compose_api.go

```diff
@@ -0,0 +1,291 @@
+	}
+
+	upOptions := stackUpOptions{
+		CreateOptions: api.CreateOptions{
+			Services:             d.project.ServiceNames(),
+			Recreate:             api.RecreateDiverged,
+			RecreateDependencies: api.RecreateDiverged,
+		},
+		StartOptions: api.StartOptions{
+			Project: d.project,
+		},
+	}
+
+	for i := range opts {
+		opts[i].applyToStackUp(&upOptions)
+	}
+
+	if len(upOptions.Services) != len(d.project.Services) {
+		sort.Strings(upOptions.Services)
+
+		filteredServices := make(types.Services, 0, len(d.project.Services))
+
+		for i := range d.project.Services {
+			if idx := sort.SearchStrings(upOptions.Services, d.project.Services[i].Name); idx < len(upOptions.Services) && upOptions.Services[idx] == d.project.Services[i].Name {
+				filteredServices = append(filteredServices, d.project.Services[i])
+			}
+		}
+
+		d.project.Services = filteredServices
+	}
+
+	err = d.composeService.Up(ctx, d.project, api.UpOptions{
+		Create: upOptions.CreateOptions,
+		Start:  upOptions.StartOptions,
+	})
+
+	if len(d.waitStrategies) == 0 || err != nil {
+		return err
+	}
+
+	errGrp, errGrpCtx := errgroup.WithContext(ctx)
+
+	for svc, strategy := range d.waitStrategies { // pinning the variables
+		svc := svc
+		strategy := strategy
+
+		errGrp.Go(func() error {
+			target, err := d.ServiceContainer(errGrpCtx, svc)
+			if err != nil {
+				return err
+			}
+			return strategy.WaitUntilReady(errGrpCtx, target)
+		})
+	}
+
+	return errGrp.Wait()
+}
+
+func (d *dockerComposeAPI) WaitForService(s string, strategy wait.Strategy) ComposeStack {
+	d.waitStrategies[s] = strategy
+	return d
+}
+
+func (d *dockerComposeAPI) WithEnv(m map[string]string) ComposeStack {
+	d.projectOptions = append(d.projectOptions, withEnv(m))
+	return d
+}
+
+func (d *dockerComposeAPI) WithOsEnv() ComposeStack {
+	d.projectOptions = append(d.projectOptions, cli.WithOsEnv)
+	return d
+}
+
+func (d *dockerComposeAPI) ServiceContainer(ctx context.Context, svcName string) (*DockerContainer, error) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	if container, ok := d.containers[svcName]; ok {
+		return container, nil
+	}
+
+	listOptions := types2.ContainerListOptions{
+		All: true,
+		Filters: filters.NewArgs(
+			filters.Arg("label", fmt.Sprintf("%s=%s", api.ProjectLabel, d.name)),
+			filters.Arg("label", fmt.Sprintf("%s=%s", api.ServiceLabel, svcName)),
+		),
+	}
+
+	containers, err := d.dockerClient.ContainerList(ctx, listOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(containers) == 0 {
+		return nil, fmt.Errorf("no container found for service name %s", svcName)
+	}
+
+	containerInstance := containers[0]
+	container := &DockerContainer{
+		ID: containerInstance.ID,
+		provider: &DockerProvider{
+			client: d.dockerClient,
+		},
+	}
+
+	d.containers[svcName] = container
+
+	return container, nil
+}
+
+func (d *dockerComposeAPI) compileProject() (*types.Project, error) {
+	projectOptions := make([]cli.ProjectOptionsFn, len(d.projectOptions), len(d.projectOptions)+2)
```

Comment:
could we explain in a comment the need of increasing the capacity by two elements?

Reply:
I'll introduce a constant with a speaking name, hope that does the trick for you as well :wink: